ocaml/xapi/xapi_pool.ml

```diff
@@ -2072,9 +2072,45 @@ let disable_ssl_legacy = set_ssl_legacy_on_each_host ~value:false
 
 let enable_ssl_legacy = set_ssl_legacy_on_each_host ~value:true
 
+let assert_igmp_snooping_licensed ~__context =
+  Pool_features.assert_enabled ~__context ~f:Features.IGMP_snooping
+
 let set_igmp_snooping_enabled ~__context ~self ~value = 
-  Db.Pool.set_igmp_snooping_enabled ~__context ~self ~value
-  (* To be filled in later. *)
+    if value = true then
+      assert_igmp_snooping_licensed ~__context;
+
+    let dbg = Context.string_of_task __context in
+    let backend = Net.Bridge.get_kind dbg () in
+    if backend <> Network_interface.Openvswitch then
+      raise (Api_errors.Server_error (Api_errors.operation_not_allowed, ["System must be configured for OVS to set IGMP Snooping"]));
+
+    Db.Pool.set_igmp_snooping_enabled ~__context ~self ~value;
+    let hosts = Db.Host.get_all ~__context in
+    let networks = Db.Network.get_all ~__context in
+
+    let igmp_networks =
+      List.filter (fun self ->
+        let pifs = Db.Network.get_PIFs ~__context ~self:self in
+        if pifs <> [] then begin            
+          let pif = List.hd(pifs) in
+          let pif_record = Db.PIF.get_record ~__context ~self:pif in
+          if (pif_record.API.pIF_VLAN = -1L) && (pif_record.API.pIF_bond_slave_of = Ref.null) then
+            true
+          else
+            false
+        end
+        else false
+      ) networks
+    in
+    List.iter (fun host ->
+      List.iter (fun network ->
+        try
+          Helpers.call_api_functions ~__context (fun rpc session_id -> 
+              Client.Network.attach ~rpc ~session_id ~network ~host)
+        with _ ->
+          debug "set_igmp_snooping_enabled:Network.attach failed on host=%s network=%s" (Ref.string_of host) (Ref.string_of network)
```

Comment:
We need to raise an error if any of the attach calls fails (but still try all and not return immediately on error). You can use `List.fold_left` instead of `List.iter`, to return an `error` Boolean so that you can raise an `Api_error` at the end.